pkg/ext/resources/tokens/store.go

```diff
@@ -0,0 +1,325 @@
+		}
+		token.Status.PlaintextToken = tokenValue
+		hashedValue, err := hashers.GetHasher().CreateHash(tokenValue)
+		if err != nil {
+			return nil, fmt.Errorf("unable to hash token value: %w", err)
+		}
+		token.Status.HashedToken = hashedValue
+	}
+	// user is trying to create a token for another user
+	if token.Spec.UserID != userInfo.GetName() {
+		authed, err := t.userHasFullPermissions(userInfo)
+		if err != nil {
+			return nil, fmt.Errorf("unable to check if user has * on tokens: %w", err)
+		}
+		if !authed {
+			return nil, fmt.Errorf("can't create token for other user %s since user %s doesn't have * on ranchertokens", userInfo.GetName(), token.Spec.UserID)
+		}
+	}
+	secret := secretFromToken(token)
+	_, err := t.secretClient.Create(secret)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create secret for token: %w", err)
+	}
+	// users don't care about the hashed value
+	token.Status.HashedToken = ""
+	return token, nil
+}
+
+func (t *TokenStore) Update(ctx context.Context, userInfo user.Info, token *RancherToken, opts *metav1.UpdateOptions) (*RancherToken, error) {
+	// user is trying to update a token for another user
+	if token.Spec.UserID != userInfo.GetName() {
+		authed, err := t.userHasFullPermissions(userInfo)
+		if err != nil {
+			return nil, fmt.Errorf("unable to check if user has * on tokens: %w", err)
+		}
+		if !authed {
+			return nil, fmt.Errorf("can't create token for other user %s since user %s doesn't have * on ranchertokens", userInfo.GetName(), token.Spec.UserID)
+		}
+	}
+	currentSecret, err := t.secretCache.Get(TokenNamespace, token.Name)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get current token %s: %w", token.Name, err)
+	}
+	currentToken := tokenFromSecret(currentSecret)
+	token.Status.HashedToken = currentToken.Status.HashedToken
+	token.Status.PlaintextToken = ""
+	secret := secretFromToken(token)
+	newSecret, err := t.secretClient.Update(secret)
+	if err != nil {
+		return nil, fmt.Errorf("unable to update token %s: %w", token.Name, err)
+	}
+	newToken := tokenFromSecret(newSecret)
+	newToken.Status.HashedToken = ""
+	newToken.Status.PlaintextToken = ""
+
+	return newToken, nil
+}
+
+func (t *TokenStore) Get(ctx context.Context, userInfo user.Info, name string, opts *metav1.GetOptions) (*RancherToken, error) {
+	currentSecret, err := t.secretCache.Get(TokenNamespace, name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		return nil, fmt.Errorf("unable to get token %s: %w", name, err)
+	}
+	token := tokenFromSecret(currentSecret)
+	// user is trying to get a token for another user
+	if token.Spec.UserID != userInfo.GetName() {
+		authed, err := t.userHasFullPermissions(userInfo)
+		if err != nil {
+			return nil, fmt.Errorf("unable to check if user has * on tokens: %w", err)
+		}
+		if !authed {
+			return nil, fmt.Errorf("can't create token for other user %s since user %s doesn't have * on ranchertokens", userInfo.GetName(), token.Spec.UserID)
```

Comment:
```suggestion
			return nil, fmt.Errorf("can't get token for other user %s since user %s doesn't have * on ranchertokens", userInfo.GetName(), token.Spec.UserID)
```

Reply:
Also, move auth/perm check before existence check! Do not leak if a specific token exists, even if we cannot manipulate it.